rfcs/006-mango-fdb.md

```diff
@@ -0,0 +1,154 @@
+
+## Data Model
+
+### Index Definitions
+
+A Mango index is defined as:
+
+```json
+{
+  name: ‘view-name’ - optional will be auto-generated
+  index: {
+    fields: [‘fieldA’, ‘fieldB’] - fields to be indexed
+  },
+  partial_filter_selector {} - optional filter to process documents before adding to the index
+}
+```
+
+The above index definition would be stored in FoundationDB as:
+
+`(?DATABASE, ?INDEX_DEFINITIONS, <fieldname1>, …<rest of fields>) = (<index_name>, <partial_filter_selector>, build_status, sequence)`
+
+`build_status`  will have two options, `active` which indicates the index is ready to service queries or `building` if the index is still being built. `sequence` is the sequence that the index is created at. Nested fields defined in the index would be stored as packed tuples.
+
+### Indexes
+
+Each index defined in the Index Definition would have an index key space where the database’s documents are stored and sorted via the keys defined in the index’s definition. The data model for each defined index would be:
+
+`(?DATABASE, ?INDEXES, ?INDEX_NAME,  <indexed_field>, …<other indexed fields>, _id) = null`
+
+The `_id` is kept to avoid duplicate keys and to be used to retrieve the full document for a Mango query.
+For now, the value will be null, later we can look at storing covering indexes. aggregate values or materialised views.
+
+### Key sorting
+
+In CouchDB 2.x ICU collation is used to sort string key’s when added to the index’s b-tree. The current way of using ICU string collation won’t work with FoundationDB. To resolve this strings will be converted to an ICU sort string before being stored in FDB.  This is an extra performance overhead but will only be done when one when writing a key into the index.
+
+CouchDB has a defined [index collation specification](http://docs.couchdb.org/en/stable/ddocs/views/collation.html#collation-specification) that the new Mango design must adhere to. Each key added to a Mango index will be converted into a composite key or tuple with the first value in the tuple representing the type that the key so that it would be sorted correctly. Below is an example of the type keys to be used:
+
+\x00 NULL
+\x26 False
+\x27 True
+\x30 Numbers
+\x40 Text converted into a sort string
+\x50 Array
+\x60 Objects
+
+An example for a number key would be (\x30, 1). Just too note, Null and Boolean values won’t need to be composite keys as the type key is the value.
+
+### Index Limits
+
+This design has certain defined limits for it to work correctly:
+
+* The index definition (name, fields and partial_filter_selector) cannot exceed 100 KB FDB value limit
+* The sorted keys for an index cannot exceed the 10 KB key limit
```

Comment:
"sorted keys" - is this the same as the keys emitted for indexing records?

Reply:
Correct. They are.